### PR TITLE
FEATURE mc_issue_set_handler()

### PR DESCRIPTION
**Pull request by fmancardi**
I've developed this (may be quick & dirty) using mc_issue_update() as starting point. My need was to have possibility to update/set _just_ issue handler (using userID), a thing that seems not simple using mc_issue_update().
Please consider to add this to stable code
Thanks a lot for your wonderful application.

### CODE REVIEW

File: api/soap/mc_issue_api.php [hunk 1]
```diff
@@ -1597,3 +1597,59 @@ function mci_issue_data_as_header_array( $p_issue_data ) {
...
+
+	$t_project_id = bug_get_field( $p_issue_id, 'project_id' );
+
+	if( !mci_has_readwrite_access( $t_user_id, $t_project_id ) ) {
+		return mci_soap_fault_access_denied( $t_user_id );
+	}
+
+	$g_project_override = $t_project_id; // ensure that helper_get_current_project() calls resolve to this project id
+
+    if( is_null($p_handler_id) && $p_handler_id <= 0) {
```
Comment (vboctor) [inline api/soap/mc_issue_api.php]: Did you mean to have an || (OR) here?   Adjust spacing around parenthesis to match rest of code and indentation.

**Comment by vboctor** [comment]
Thanks @fmancardi for your contribution.

I suggest we consider the following:
- Rename field to mc_issue_assign().
- Send handler as AccountData similar to mc_issue_add() enable clients to provide username, id, or email.
- Should we enable an optional IssueNoteData to provide optional addition of a note as part of assignment.

It also shouldn't be too hard to achieve the same via an mc_issue_get() and mc_issue_update(), where the mc_issue_update() if the last update field is provided in the incoming request, yet it doesn't match the one in the data (i.e. someone else update the bug between the get and the update).  However, I can understand the need for a more specific API that requires a single round trip.

@rombert can also provide feedback on this PR.

**Comment by fmancardi** [comment]
First thanks for your fat answer
1) Yes, I was wrong when using AND, instead was OR; excuse me
2) I've thought about using AccountData as input, OK I will change do all changes in method interface and name that you have suggested.

3) Before developing this I've given a look to the mc_issue_update() because I suppose it can do the trick, but I've discovered that this method requires lot of mandatory fields when you only want to do the assignment, then IMHO seems an overkill. I think we can still use you approach i.e. rewriting this method using mc_issue_get() and mc_issue_update() in this way:
using p_issue_id => use mc_issue_get() to get issue data, then on the object you got, set ONLY handler and notes (is any exists), then call mc_issue_update().
Do you prefer this approach?

4) I can change the method name, no problems.
    Only issue is that I do not how to proceed: i.e. I need to open a new merge request ? or this can be change?
Can I ask via mail to rmobert?

5) Do you want the opposite method mc_issue_unassign() we written? this is neede if we want to have same feature present on GUI

Best regards
Francisco

**Comment by dregad** [comment]
> 4) I can change the method name, no problems.
Only issue is that I do not how to proceed: i.e. I need to open a new merge request ? or this can be change?

Just add new commits to your local branch and push them to your github fork, they will appear here automatically. In the future, I suggest you create a local feature branch instead of updating "master-1.2.x".

**Comment by rombert** [comment]
> Before developing this I've given a look to the mc_issue_update() because I suppose it can do the trick, but I've discovered that this method requires lot of mandatory fields when you only want to do the assignment, then IMHO seems an overkill. 

If I read the code for `mc_issue_update` correctly ( https://github.com/mantisbt/mantisbt/blob/8e35696dd1d097696ca8532a96eef245964ce195/api/soap/mc_issue_api.php#L931-L944 ) then the following fields are required:

* project_id
* reporter_id
* handler_id
* category_id
* summary
* description

Actually, I think that no field should be required for updates, so IMO a better approach would be to change the `mc_issue_update` method to not require any fields to be set in the call, so it's easier to change just one field.

I understand why it's simpler to add a `mc_issue_set_handler` method, but then again the same argument can be made for (almost) any field that can be changed. I'd rather see `mc_issue_update` being improved.

**Comment by vboctor** [comment]
There are a couple of things to take into consideration with the update method:
- There needs to be a distinction between "not supplied" and "not set".  For example, in an update request, handler_id can be not supplied or set to something that unassigns the issue.
- We should fail update requests that are based on an older version of the issue if the date_modified field is supplied, but doesn't match the database.

**Comment by fmancardi** [comment]
IHMO there are other things to think about
If the target of API is to do things with same constraints that actions
done by a user on GUI then maybe some issues attributes related to audit
like date submitted, datemodified may be need to be ignored by API.
if API allows setting properties that can not be normally managed on GUI
this means that using API is like accessing DB using a SQL client.
It's not a black or white choice (IMHO) because to integrate system YOU
want to be able to set the reporter and normally when using GUI you can not
do this.
Hope i was clear enough.

Regards
Il 20/Nov/2015 04:56, "Victor Boctor" <notifications@github.com> ha scritto:

**Comment by dregad** [comment]
What is the status of this PR ?

**Comment by l266509** [comment]
@vboctor @giallu @daryn @rombert @atrol @grangeway @dregad @Dentxinho @bieli @tomcarlson @syncguru @diegovice @caribe @fager @nikoutel @6RUN0 @AgnidykhBayky @sayyidsh @PixelstreamAU @JohnMS @mikeching @g-taki @benv8nam @stew721 @sayazyi @lezouretalice @SuriyaaKudoIsc @solovets @kacmem @xiaojiangang @chenyongze @myu9882 @markmarco16 @mgretler @lincolnwung @krime @scottmatheina @FavioGalvis @vand3rson @like9729 @badfiles @zenithtekla @SureShinde @netsf @csatwood @miao606 @mdj20130907 @bitlin @ioroid @supertassu @TasCraft @fernandodelucca @ulkuderner @vbathke @hersonpc @jrtorres42 @thejuanwhocode @cyuankai @mantianxingxing @shirleycharlin @boby75 @Lordshell @BurningWind @jaguarxii @Praymundo @johnzastrow @ewmartinez @marcioklein @lypzln @nospam1961 @jllano @lincoln56 @modir @xiaocnie @sivpro @planetbrett @teodortalov @ltpgt @pbcreativesolutions @Gantha @Gerolkae @gunsmoker @nikerabbit @translatewiki  Whats status of this PR?

**Pull request closed without merging**